File: 90-subsets-ii/subsets-ii.cpp
class Solution {
public:
    // Global
    vector<vector<int>> ans;

    void solve(vector<int> &nums, vector<int> &temp, int index){
        if (index == nums.size()){
            ans.push_back(temp);
            return ;
        }

        temp.push_back(nums[index]);
        solve(nums, temp, index + 1);
        temp.pop_back();

        while(index + 1 < nums.size() && nums[index] == nums[index+1]){
            index += 1;
        }

        solve(nums, temp, index+1);
    }
    vector<vector<int>> subsetsWithDup(vector<int>& nums) {
        vector<int> temp;
        sort(nums.begin(), nums.end());
        solve(nums, temp, 0);
        return ans;
    }
};

File: 90-subsets-ii/subsets-ii.go
func solve(nums []int, idx int, temp []int, ans *[][]int) {
    tmp := make([]int, len(temp))
    copy(tmp, temp)
    *ans = append(*ans, tmp)

    for i := idx; i < len(nums); i++ {
        if i > idx && nums[i] == nums[i-1] {
            continue
        }

        temp = append(temp, nums[i])
        solve(nums, i+1, temp, ans)
        temp = temp[:len(temp)-1]
    }
}

func subsetsWithDup(nums []int) [][]int {
    sort.Ints(nums)

    var temp []int
    var ans [][]int

    solve(nums, 0, temp, &ans)

    return ans
}